### PR TITLE
CASSANDRA-20460: Expand TableWalk tests to include collections and add support for += and -= for these types

### PR DESCRIPTION
**Pull request by dcapwell**

### CODE REVIEW

File: test/distributed/org/apache/cassandra/distributed/test/cql3/SingleNodeTableWalkTest.java [hunk 1]
```diff
@@ -88,10 +101,20 @@ protected void preCheck(Cluster cluster, Property.StatefulBuilder builder)
...
 
-    protected TypeGenBuilder supportedTypes()
+    protected TypeGenBuilder supportedTypes(RandomSource rs)
     {
-        return AbstractTypeGenerators.withoutUnsafeEquality(AbstractTypeGenerators.builder()
-                                                                                  .withTypeKinds(AbstractTypeGenerators.TypeKind.PRIMITIVE));
+        return AbstractTypeGenerators.builder()
+                                     .withTypeKinds(Generators.fromGen(TYPE_KIND_DISTRIBUTION.next(rs)))
+                                     .withPrimitives(Generators.fromGen(PRIMITIVE_DISTRIBUTION.next(rs)))
+                                     .withUserTypeFields(AbstractTypeGenerators.UserTypeFieldsGen.simpleNames())
```
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/test/cql3/SingleNodeTableWalkTest.java:108]: not testing UDT serialization so simple names makes it so much easier to read...

File: test/distributed/org/apache/cassandra/distributed/test/cql3/SingleNodeTableWalkTest.java [hunk 2]
```diff
@@ -88,10 +101,20 @@ protected void preCheck(Cluster cluster, Property.StatefulBuilder builder)
...
     }
 
-    protected TypeGenBuilder supportedTypes()
+    protected TypeGenBuilder supportedTypes(RandomSource rs)
     {
-        return AbstractTypeGenerators.withoutUnsafeEquality(AbstractTypeGenerators.builder()
-                                                                                  .withTypeKinds(AbstractTypeGenerators.TypeKind.PRIMITIVE));
+        return AbstractTypeGenerators.builder()
+                                     .withTypeKinds(Generators.fromGen(TYPE_KIND_DISTRIBUTION.next(rs)))
+                                     .withPrimitives(Generators.fromGen(PRIMITIVE_DISTRIBUTION.next(rs)))
```
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/test/cql3/SingleNodeTableWalkTest.java:107]: before we used uniform random so each test had the same weights.  Now we select a new distribution for each test run (some tests will only be primitives, others will be mostly maps, etc.)

File: test/distributed/org/apache/cassandra/distributed/test/cql3/SingleNodeTableWalkTest.java [hunk 3]
```diff
@@ -88,10 +101,20 @@ protected void preCheck(Cluster cluster, Property.StatefulBuilder builder)
...
+                                     .withUserTypeFields(AbstractTypeGenerators.UserTypeFieldsGen.simpleNames())
+                                     .withMaxDepth(1);
```
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/test/cql3/SingleNodeTableWalkTest.java:109]: ATM we are not testing nested access (will need to eventually) so lowered as the cql literal output was so large it got hard to read (UDT of map of UDT of map of UDT....)

File: test/distributed/org/apache/cassandra/distributed/test/cql3/SingleNodeTableWalkTest.java [hunk 4]
```diff
@@ -206,7 +229,7 @@ protected List<CreateIndexDDL.Indexer> supportedIndexers()
             builder.value(pk, key.bufferAt(pks.indexOf(pk)));
 
 
-        List<Symbol> searchableColumns = state.nonPartitionColumns;
+        List<Symbol> searchableColumns = state.searchableNonPartitionColumns;
```
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/test/cql3/SingleNodeTableWalkTest.java:231]: multi cell collections/udt can not be searched, so just using the non-partition columns wasn't enough

File: test/distributed/org/apache/cassandra/distributed/test/cql3/SingleNodeTableWalkTest.java [hunk 5]
```diff
@@ -363,7 +386,8 @@ protected TableMetadata defineTable(RandomSource rs, String ks)
                                        .withCompression())
                      .withKeyspaceName(ks).withTableName("tbl")
                      .withSimpleColumnNames()
-                     .withDefaultTypeGen(supportedTypes())
+                     .withDefaultTypeGen(supportedTypes(rs))
+                     .withPrimaryColumnTypeGen(supportedPrimaryColumnTypes(rs))
```
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/test/cql3/SingleNodeTableWalkTest.java:389]: when primary keys can be collections there can be issues with having too many bytes and having our mutations be rejected.  Keeping it primitive also improves readability for debug output...

Would be good to have a test that overrides this though, just to improve coverage...

File: test/distributed/org/apache/cassandra/distributed/test/cql3/SingleNodeTableWalkTest.java [hunk 6]
```diff
@@ -424,7 +448,8 @@ public State(RandomSource rs, Cluster cluster)
                                                                   .withoutTransaction()
                                                                   .withoutTtl()
                                                                   .withoutTimestamp()
-                                                                  .withPartitions(SourceDSL.arbitrary().pick(uniquePartitions));
+                                                                  .withPartitions(SourceDSL.arbitrary().pick(uniquePartitions))
+                                                                  .withColumnExpressions(e -> e.withOperators(Generators.fromGen(BOOLEAN_DISTRIBUTION.next(rs))));
```
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/test/cql3/SingleNodeTableWalkTest.java:451]: each column will have a different distribution for when operators are allowed

File: test/harry/main/org/apache/cassandra/harry/model/ASTSingleTableModel.java [hunk 7]
```diff
@@ -493,6 +503,45 @@ private static void validateAnyOrder(ImmutableUniqueList<Symbol> columns, Set<Ro
             if (actual.isEmpty()) sb.append("No rows returned");
             else sb.append("Missing rows:\n").append(table(columns, missing));
         }
+        if (!unexpected.isEmpty() && unexpected.size() == missing.size())
```
Comment (dcapwell) [inline test/harry/main/org/apache/cassandra/harry/model/ASTSingleTableModel.java:506]: this is extra human readable validation logic.  `null` vs `<empty>` has been super hard to see from the above output, and the use case i was finding is that we had the correct rows returned, but 1 or more columns had a difference... so if this is the case we try to show which columns are off... this was super useful for detecting consistency issues with `null` and `<empty>`

File: test/harry/main/org/apache/cassandra/harry/model/ASTSingleTableModel.java [hunk 8]
```diff
@@ -493,6 +503,45 @@ private static void validateAnyOrder(ImmutableUniqueList<Symbol> columns, Set<Ro
...
+        {
+            // good chance a column differs
+            StringBuilder finalSb = sb;
+            Runnable runOnce = new Runnable()
```
Comment (dcapwell) [inline test/harry/main/org/apache/cassandra/harry/model/ASTSingleTableModel.java:510]: i think Alex created a reusable function for this in accord... would need to find it

File: test/harry/main/org/apache/cassandra/harry/model/ASTSingleTableModel.java [hunk 9]
```diff
@@ -731,7 +780,7 @@ private static boolean matches(ByteBuffer value, List<? extends Expression> cond
         for (Expression e : conditions)
         {
             ByteBuffer expected = eval(e);
-            if (expected.equals(value))
+            if (expected != null && expected.equals(value))
```
Comment (dcapwell) [inline test/harry/main/org/apache/cassandra/harry/model/ASTSingleTableModel.java:783]: `eval` is now `Nullable` but this code path shouldn't really ever see a null... maybe if you did something like `[1] - [1]` then we could get a `null` here...  mostly added to be correct and defensive (and intellij yelled at me)

File: test/harry/main/org/apache/cassandra/harry/model/BytesPartitionState.java [hunk 10]
```diff
@@ -135,7 +141,13 @@ private long[] toDescriptor(ImmutableUniqueList<Symbol> positions, Map<Symbol, B
             Symbol column = positions.get(i);
             if (values.containsKey(column))
             {
-                long vd = factory.valueCache.deflate(new Value(column.type(), values.get(column)));
+                ByteBuffer value = values.get(column);
+                if (value == null || !value.hasRemaining() && (column.type().isUDT() && column.type().isMultiCell()))
```
Comment (dcapwell) [inline test/harry/main/org/apache/cassandra/harry/model/BytesPartitionState.java:146]: user type is the only multi cell type i see that allows `<empty>` so this check should be fine; can expand if we find more cases

File: test/unit/org/apache/cassandra/cql3/ast/CreateIndexDDL.java [hunk 11]
```diff
@@ -141,7 +141,7 @@ else if (((isFrozen(type) && !type.isVector()) || StorageAttachedIndex.SUPPORTED
         public EnumSet<QueryType> supportedQueries(AbstractType<?> type)
         {
             type = type.unwrap();
-            if (IndexTermType.isEqOnlyType(type))
+            if (IndexTermType.isEqOnlyType(type) || type.isCollection() || type.isUDT() || type.isTuple())
```
Comment (dcapwell) [inline test/unit/org/apache/cassandra/cql3/ast/CreateIndexDDL.java:144]: @maedhroz should we push this into `IndexTermType.isEqOnlyType`?

File: test/unit/org/apache/cassandra/cql3/ast/ExpressionEvaluator.java [hunk 12]
```diff
@@ -21,126 +21,152 @@
...
+import javax.annotation.Nullable;
 
-import static java.util.Optional.of;
+import org.apache.cassandra.db.marshal.AbstractType;
 
 public class ExpressionEvaluator
 {
-    public static Optional<Object> tryEval(Expression e)
+    @Nullable
+    public static Object eval(Expression e)
```
Comment (dcapwell) [inline test/unit/org/apache/cassandra/cql3/ast/ExpressionEvaluator.java:38]: returning `Optional` was just causing confusion, and all code paths did `.get()` anyways.

I also needed the ability to return `null` rather than `empty` which i couldn't do with `Optional`

File: test/unit/org/apache/cassandra/cql3/ast/ExpressionEvaluator.java [hunk 13]
```diff
@@ -21,126 +21,152 @@
...
-        return Optional.empty();
+            return eval((Operator) e);
+        throw new UnsupportedOperationException("Unexpected expression type " + e.getClass() + ": " + e.toCQL());
     }
 
-    public static Optional<Object> tryEval(Operator e)
+    @Nullable
+    public static Object eval(Operator e)
     {
+        Object lhs = eval(e.left);
```
Comment (dcapwell) [inline test/unit/org/apache/cassandra/cql3/ast/ExpressionEvaluator.java:52]: the model `eval` method will put the "current" value as a `ByteBuffer` into the operation, this then means that the types might not match!  to solve this we just convert w/e we see a BB.

File: test/unit/org/apache/cassandra/cql3/ast/ExpressionEvaluator.java [hunk 14]
```diff
@@ -21,126 +21,152 @@
...
+                    accum.removeAll((Set<Object>) rhs);
+                    return accum.isEmpty() ? null : accum;
+                }
+                if (lhs instanceof List)
+                {
+                    List<Object> accum = new ArrayList<>((List<Object>) lhs);
+                    accum.removeAll((List<Object>) rhs);
+                    return accum.isEmpty() ? null : accum;
+                }
+                if (lhs instanceof Map)
```
Comment (dcapwell) [inline test/unit/org/apache/cassandra/cql3/ast/ExpressionEvaluator.java:130]: this code path is currently untested, only part of this patch that wasn't validated

File: test/unit/org/apache/cassandra/utils/ASTGenerators.java [hunk 15]
```diff
@@ -770,12 +784,12 @@ private void generateRemaining(RandomnessSource rnd,
                     }
                 }
             }
-            if (kind == Mutation.Kind.UPDATE && isTransaction)
```
Comment (dcapwell) [inline test/unit/org/apache/cassandra/utils/ASTGenerators.java:773]: collections can do things outside of transactions, so need to push this into `supportsOperators`

File: test/unit/org/apache/cassandra/utils/AbstractTypeGenerators.java [hunk 16]
```diff
@@ -289,19 +298,27 @@ public TypeGenBuilder()
...
+            compositeSizeGen = other.compositeSizeGen;
             primitiveGen = other.primitiveGen;
+            compositeElementGen = other.compositeElementGen;
             userTypeKeyspaceGen = other.userTypeKeyspaceGen;
             defaultSetKeyFunc = other.defaultSetKeyFunc;
-            compositeElementGen = other.compositeElementGen;
-            compositeSizeGen = other.compositeSizeGen;
             typeFilter = other.typeFilter;
+            udtName = other.udtName;
+            multiCellGen = other.multiCellGen;
```
Comment (dcapwell) [inline test/unit/org/apache/cassandra/utils/AbstractTypeGenerators.java:314]: this was missing so i took the fields in object order and recreated this constructor just in case there was more; looks like it was just this one though.

File: test/unit/org/apache/cassandra/utils/Generators.java [hunk 17]
```diff
@@ -492,13 +494,20 @@ public static <T> Gen<Set<T>> set(Gen<T> gen, Gen<Integer> sizeGen)
...
-            return list;
-        });
+        return rnd -> {
+            int size = sizeGen.generate(rnd);
+            Set<T> set = Sets.newHashSetWithExpectedSize(size);
+            List<T> output = new ArrayList<>(size);
+            for (int i = 0; i < size; i++)
+            {
+                T value;
+                while (!set.add(value = gen.generate(rnd))) {}
```
Comment (dcapwell) [inline test/unit/org/apache/cassandra/utils/Generators.java:506]: this isn't as safe as `uniqueBestEffort` from accord (if gen is a boolean gen and size is 3, this will loop forever)... once accord merges we can talk about unifying all this stuff...

File: test/harry/main/org/apache/cassandra/harry/model/ASTSingleTableModel.java [hunk 18]
```diff
@@ -893,28 +942,60 @@ private List<Clustering<ByteBuffer>> keys(Map<Symbol, List<? extends Expression>
         return current.stream().map(BufferClustering::new).collect(Collectors.toList());
     }
 
+    private static ByteBuffer eval(Symbol col, @Nullable ByteBuffer current, Expression e)
+    {
+        if (!(e instanceof AssignmentOperator)) return eval(e);
+        // multi cell collections have the property that they do update even if the current value is null
+        boolean isFancy = col.type().isCollection() && col.type().isMultiCell();
```
Comment (bbotella) [inline test/harry/main/org/apache/cassandra/harry/model/ASTSingleTableModel.java:949]: Curious about the `isFancy` name choosing :-)

File: test/unit/org/apache/cassandra/cql3/ast/AssignmentOperator.java [hunk 19]
```diff
@@ -53,31 +53,30 @@ public AssignmentOperator(Kind kind, Expression right)
...
     {
         type = type.unwrap();
         EnumSet<Kind> result = EnumSet.noneOf(Kind.class);
+        if (type instanceof CollectionType && type.isMultiCell())
+        {
+            if (type instanceof SetType || type instanceof ListType)
+                return EnumSet.of(Kind.ADD, Kind.SUBTRACT);
+            if (type instanceof MapType)
+            {
+                // map supports subtract, but not map - map; only map - set!
```
Comment (bbotella) [inline test/unit/org/apache/cassandra/cql3/ast/AssignmentOperator.java:66]: That's... an interesting finding. Should we have a Jira to add proper support?

Comment (dcapwell) [reply test/unit/org/apache/cassandra/cql3/ast/AssignmentOperator.java:66]: i wouldn't add a JIRA unless someone wants that feature.  `map -=` acts on the key level, which can be done without a transaction (CAS or Accord), where as `remove(key, value)` would need to know the current value, which isn't behavior we like to support outside of transactions.

File: test/harry/main/org/apache/cassandra/harry/model/ASTSingleTableModel.java (hunk 18 above)
Comment (dcapwell) [reply test/harry/main/org/apache/cassandra/harry/model/ASTSingleTableModel.java:949]: i felt it was fancy!

File: test/distributed/org/apache/cassandra/distributed/test/cql3/SingleNodeTableWalkTest.java [hunk 20]
```diff
@@ -523,6 +561,8 @@ private List<Symbol> multiColumnQueryColumns()
             List<Symbol> allowedColumns = searchableColumns;
             if (hasMultiNodeMultiColumnAllowFilteringWithLocalWritesIssue())
                 allowedColumns = nonPkIndexedColumns;
+            if (IGNORED_ISSUES.contains(KnownIssue.SAI_AND_VECTOR_COLUMNS) && !indexes.isEmpty())
```
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/test/cql3/SingleNodeTableWalkTest.java:564]: added to get clean CI.  Something is up with multi column queries that touch both SAI and `vector` type.  It doesn't look like the `vector` must be indexed to get incorrect results, so doesn't feel like we are selecting a bad index here

File: test/distributed/org/apache/cassandra/distributed/test/cql3/SingleNodeTableWalkTest.java [hunk 21]
```diff
@@ -78,6 +79,17 @@
...
+    //TODO (coverage): COMPOSITE, DYNAMIC_COMPOSITE
+    private static Gen<Gen<TypeKind>> TYPE_KIND_DISTRIBUTION = Gens.mixedDistribution(TypeKind.PRIMITIVE,
+                                                                                      TypeKind.SET, TypeKind.LIST, TypeKind.MAP,
+                                                                                      TypeKind.TUPLE, TypeKind.UDT,
+                                                                                      TypeKind.VECTOR
+    );
+    private static Gen<Gen<AbstractType<?>>> PRIMITIVE_DISTRIBUTION = Gens.mixedDistribution(AbstractTypeGenerators.knownPrimitiveTypes()
+                                                                                                                   .stream()
+                                                                                                                   .filter(t -> !AbstractTypeGenerators.isUnsafeEquality(t))
+                                                                                                                   .collect(Collectors.toList()));
```
Comment (aratno) [inline test/distributed/org/apache/cassandra/distributed/test/cql3/SingleNodeTableWalkTest.java]: Can be final

File: test/unit/org/apache/cassandra/utils/Generators.java (hunk 17 above)
Comment (aratno) [reply test/unit/org/apache/cassandra/utils/Generators.java:506]: In the meantime, want to set an iteration limit with a specific error? Just to make it a bit easier to figure out what's happening if a test is timing out.

File: test/harry/main/org/apache/cassandra/harry/model/BytesPartitionState.java (hunk 10 above)
Comment (aratno) [reply test/harry/main/org/apache/cassandra/harry/model/BytesPartitionState.java:146]: Can you add this as a comment inline?

File: test/distributed/org/apache/cassandra/distributed/test/cql3/SingleNodeTableWalkTest.java (hunk 6 above)
Comment (aratno) [reply test/distributed/org/apache/cassandra/distributed/test/cql3/SingleNodeTableWalkTest.java:451]: Required for CASSANDRA-20449 repro at least 😄

**Review by aratno — APPROVED** [review]

File: test/distributed/org/apache/cassandra/distributed/test/cql3/SingleNodeTableWalkTest.java (hunk 6 above)
Comment (dcapwell) [reply test/distributed/org/apache/cassandra/distributed/test/cql3/SingleNodeTableWalkTest.java:451]: wasn't required; we hit the issue before this change.  This change is to better cover different edge cases

File: test/unit/org/apache/cassandra/utils/Generators.java (hunk 17 above)
Comment (dcapwell) [reply test/unit/org/apache/cassandra/utils/Generators.java:506]: for now no.  There are better impl in `Gens` and when i migrate off of quick theories we can just use the better versions; so no reason to handle here...

File: test/distributed/org/apache/cassandra/distributed/test/cql3/SingleNodeTableWalkTest.java (hunk 6 above)
Comment (aratno) [reply test/distributed/org/apache/cassandra/distributed/test/cql3/SingleNodeTableWalkTest.java:451]: We need different assignment operators to repro (`+= v` vs. `= {v}`) but not necessarily different distributions of operators, which is what's introduced here - is that right?

**Comment by dcapwell** [comment]
merged
